### ISSUE TITLE
Simple vectorization for batches of bounded-depth MCTSs

### ISSUE THREAD

**Issue by Philip-Bachman** [Mononofu/mcts_benchmarks#1]
You can get a nice speed boost for pure Python MCTS by vectorizing the logic for running N tree searches in parallel. The basic idea is to store the data describing, e.g., parent and child pointers describing the N trees in pre-allocated numpy/torch arrays and to run the current simulation for each of the N searches in lock step. Simulations can walk the trees in parallel using gather ops based on indices pulled from the arrays of child/parent pointers. Actions are selected via argmax over the appropriate axis of an array storing per-action scores for each tree node. You can handle variable length of the current simulation (for the N trees we're growing in parallel) by introducing a "null" node that acts as an absorbing state for the tree walking procedure that finds which node the current simluation will expand in each of the N trees.

**Comment by Mononofu** [comment]
Thanks for the suggestion!

I considered this option, but was a bit wary that it might make the core
logic hard to read and modify. My intention was to investigate how
performant a relatively pseudocode-like MCTS could be in Python, to make it
easy to test modifications to the core algorithm.

Did you have good experiences with batched MCTS?

**Comment by Philip-Bachman** [comment]
Yeah, the batched/vectorized MCTS was helpful for a project I'm currently involved in. We're working to reproduce some parts of MuZero and the initial implementation was bottlenecking on the tree walking part of the MCTS simulations, rather than on the dynamics model evaluations as one would expect. Vectorizing the MCTS shifted the bottleneck to the dynamics model evaluations.

I agree that the core tree search logic gets a bit lost in the vectorized version. Though, with some nice comments it makes a nice example of non-trivial vectorization.

We started with a pure Python port of the pseudocode you provided for the paper, so thanks for that.
